core/src/scene/importer.cpp

```diff
@@ -12,118 +13,127 @@ using YAML::NodeType;
     std::vector<Url> sceneStack;
-    importScenesRecursive(platform, scene, root, rootScenePath, sceneStack);
+    importScenesRecursive(root, sceneUrl, sceneStack);
 
     return root;
 }
 
-void Importer::processScene(const std::shared_ptr<Platform>& platform, std::shared_ptr<Scene>& scene,
-                            const Url& scenePath, const std::string &sceneString) {
+void Importer::addSceneData(const Url& sceneUrl, std::vector<char>& sceneContent) {
 
-    LOGD("Process: '%s'", scenePath.string().c_str());
+    LOGD("Process: '%s'", sceneUrl.string().c_str());
 
     // Don't load imports twice
-    if (m_scenes.find(scenePath) != m_scenes.end()) {
+    if (m_importedScenes.find(sceneUrl) != m_importedScenes.end()) {
         return;
     }
 
-    try {
-        auto sceneNode = YAML::Load(sceneString);
+    Url baseUrl = sceneUrl;
+    std::string sceneString;
+    if (Url::getPathExtension(sceneUrl.path()) == "zip") {
+        // We're loading a scene from a zip archive!
+        // First, create an archive from the data.
+        auto zipArchive = std::make_shared<ZipArchive>();
+        zipArchive->loadFromMemory(sceneContent);
+        // Find the "base" scene file in the archive entries.
+        for (const auto& entry : zipArchive->entries()) {
+            auto ext = Url::getPathExtension(entry.path);
+            // The "base" scene file must have extension "yaml" or "yml" and be
+            // at the root directory of the archive (i.e. no '/' in path).
+            if ((ext == "yaml" || ext == "yml") && entry.path.find('/') == std::string::npos) {
+                // Found the base, now create a URL to resolve the contents.
+                auto encodedSourceUrl = Url::escapeReservedCharacters(sceneUrl.string());
+                baseUrl = Url("zip://" +  encodedSourceUrl);
+                // And extract the contents to the scene string.
+                sceneString.resize(entry.uncompressedSize);
+                zipArchive->decompressEntry(&entry, &sceneString[0]);
+                break;
+            }
+        }
+        // Add the archive to the scene.
+        m_scene->addZipArchive(sceneUrl, zipArchive);
+    } else {
+        sceneString = std::string(sceneContent.data(), sceneContent.size());
+    }
 
-        m_scenes[scenePath] = sceneNode;
+    addSceneString(baseUrl, sceneUrl, sceneString);
+}
 
-        for (const auto& import : getResolvedImportUrls(platform, scene, sceneNode, scenePath)) {
-            m_sceneQueue.push_back(import);
-        }
+void Importer::addSceneString(const Url& baseUrl, const Url& sceneUrl, const std::string& sceneString) {
+    Node sceneNode;
+    try {
+        sceneNode = YAML::Load(sceneString);
     } catch (YAML::ParserException e) {
         LOGE("Parsing scene config '%s'", e.what());
+        return;
+    }
+
+    m_importedScenes[sceneUrl] = sceneNode;
+
+    for (const auto& import : getResolvedImportUrls(sceneNode, baseUrl)) {
```

Comment:
I had the same idea! I tried it while working on this branch and unfortunately resolving the URLs in a scene before merging the imports causes an error for a class of use cases. Imagine a scene `a.yml` that imports `b.yml`. In `b.yml` there is a `textures` entry named `icons`. The author of `a.yml` knows that this texture exists and they use it in the definition for a style with `texture: icons`. Now, if we resolve the URLs in `a.yml` before its imports are merged in, we would have to assume that `icons` is a relative URL (since we don't know a texture by that name) and we would resolve it (incorrectly!) as `/path/to/scene_a/icons`. However if we merge `b.yml` into the scene first then we'll know that `icons` is a named texture and treat it correctly. Thankfully, this was already covered in a test case: https://github.com/tangrams/tangram-es/blob/master/tests/unit/sceneImportTests.cpp#L201

Reply:
Ohh yes does make sense. I think I did run into this, when I originally wrote this and thanks to @hjanetzek to adding a test case for this 👍